sky/serve/autoscalers.py

```diff
@@ -192,49 +234,203 @@ def evaluate_scaling(
+    """SpotOnDemandRequestRateAutoscaler: Use spot/on-demand mixture
+    to autoscale based on request rate.
+    """
+
+    NAME: Optional[str] = 'SpotOnDemandRequestRateAutoscaler'
+
+    def __init__(self, spec: 'service_spec.SkyServiceSpec') -> None:
+        super().__init__(spec)
+        self.spot_placer: 'spot_policy.SpotPlacer' = (
+            spot_policy.SpotPlacer.from_spec(spec))
+        self.extra_on_demand_replicas: int = (
+            spec.extra_on_demand_replicas if spec.extra_on_demand_replicas
+            is not None else constants.AUTOSCALER_DEFAULT_MIN_ON_DEMAND)
+
+    def _get_on_demand_resources_override_dict(self) -> Dict[str, Any]:
+        return {'use_spot': False, 'spot_recovery': None}
+
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        launched_replica_infos = [
+            info for info in replica_infos if info.is_launched
+        ]
+        self.target_num_replicas = self._get_desired_num_replicas()
+
+        num_launched_spot, num_ready_spot = 0, 0
+        num_alive_on_demand, num_ready_on_demand = 0, 0
+        for info in launched_replica_infos:
+            if info.is_spot:
+                if info.status == serve_state.ReplicaStatus.READY:
+                    num_ready_spot += 1
+                num_launched_spot += 1
+            else:
+                if info.status == serve_state.ReplicaStatus.READY:
+                    num_ready_on_demand += 1
+                num_alive_on_demand += 1
+        logger.info(
+            f'Number of alive spot instances: {num_launched_spot}, '
+            f'Number of ready spot instances: {num_ready_spot}, '
+            f'Number of alive on-demand instances: {num_alive_on_demand}, '
+            f'Number of ready on-demand instances: {num_ready_on_demand}')
+
+        scaling_options = (self._scale_spot_instances(num_launched_spot,
+                                                      launched_replica_infos))
+
+        num_to_provision = (self.target_num_replicas + self.num_overprovision)
+        num_demand_to_scale_up, num_demand_to_scale_down = 0, 0
+        if num_alive_on_demand < self.extra_on_demand_replicas:
+            num_demand_to_scale_up = (self.extra_on_demand_replicas -
+                                      num_alive_on_demand)
+        if num_ready_spot + num_alive_on_demand < num_to_provision:
+            num_demand_to_scale_up = max(
+                num_demand_to_scale_up,
+                (min(self.target_num_replicas,
+                     num_to_provision - num_ready_spot) - num_alive_on_demand))
+
+        elif (num_ready_spot + num_alive_on_demand > num_to_provision and
+              num_alive_on_demand > self.extra_on_demand_replicas):
+            num_demand_to_scale_down = (num_ready_spot + num_alive_on_demand -
+                                        num_to_provision)
```

Comment:
Could we keep track of which replicas are the extra ondemand replicas and manually separate them into 2 groups, and generate autoscaling options separately for better readability?